issuer_controller/src/issuer.py

```diff
@@ -289,6 +322,52 @@ def tob_connection_synced():
     )
 
 
+def tob_connection_active():
+    """
+    Return True if there are pending credential requests, False otherwise.
+    Note this will return False if the TOB connection is not yet sync'ed.
+    """
+    if not tob_connection_synced():
+        return False
+    return (0 < len(list(credential_requests.keys())))
+
+
+def issuer_liveness_check():
+    """
+    Check if we can shut down the container - if we have received a shutdown request and there are
+    no outstanding credential requests.
+    """
+    global app_config
+
+    if app_config["running"]:
+        # return True until we get a shutdown request
+        return True
+
+    # return True until the work queue is cleared
+    return tob_connection_active()
+
+
+class ShutdownProcessingThread(threading.Thread):
+    def run(self):
+        while issuer_liveness_check():
+            LOGGER.error("... Waiting for work queue to clear before shutdown ...")
+            time.sleep(1)
+
+
+def signal_issuer_shutdown(signum, frame):
+    """
+    Tell the issuer to do a clean shutdown (finish work queue first).
+    """
+    global app_config
+
+    LOGGER.error(">>> Received shutdown signal!")
```

Comment:
Would these better be `warning` level messages? Or the idea is to have them show on the console regardless of which log level is configured? It just seems it could be misleading to see an error in the console, while it is an expected behaviour.